### PR TITLE
r1cs: constraint_system, linear_combination: extract weights from circuit

### PR DESCRIPTION
**Pull request by akirillo**
This exposes functionality over the `ConstraintSystem` trait and `LinearCombination` type to extract the circuit weights in sparse-reduced form

### CODE REVIEW

File: tests/r1cs.rs [hunk 1]
```diff
@@ -347,20 +349,270 @@ fn example_gadget_roundtrip_serialization_helper(
...
+    // constraints[1] = (v[2] + v[3]) - a_R[0]
+
+    // o = a3 * b3
+    // a_O[0] = o
+
+    // constraints[2] = (v[4] + c2) - a_O[0]
+
+    // Thus, matrices should be:
+
+    /*
```
Comment (joeykraut) [inline tests/r1cs.rs]: Can we fix the formatting on this comment? Hard to understand what's going on

File: src/r1cs/constraint_system.rs [hunk 2]
```diff
@@ -1,8 +1,56 @@
...
+        let a: HashSet<&(usize, Scalar)> = self.0.iter().collect();
+        let b: HashSet<&(usize, Scalar)> = other.0.iter().collect();
+
+        a == b
+    }
+}
+impl Eq for SparseWeightRow {}
+
+impl Extend<(usize, Option<Scalar>)> for SparseWeightRow {
+    fn extend<T: IntoIterator<Item = (usize, Option<Scalar>)>>(&mut self, iter: T) {
```
Comment (joeykraut) [inline src/r1cs/constraint_system.rs:34]: Why is this an `Option`? Can we add a comment? Or better wrap this in a named struct `SparseRowEntry` to make the code more readable

File: src/r1cs/linear_combination.rs [hunk 3]
```diff
@@ -129,6 +131,53 @@ impl LinearCombination {
...
+        // and a single constant
+
+        let mut w_l_row = SparseWeightRow(Vec::new());
+        let mut w_r_row = SparseWeightRow(Vec::new());
+        let mut w_o_row = SparseWeightRow(Vec::new());
+        let mut w_v_row = SparseWeightRow(Vec::new());
+        let mut c = None;
+
+        for (&var, &coeff) in &self.terms {
+            if coeff != Scalar::zero() {
```
Comment (joeykraut) [inline src/r1cs/linear_combination.rs]: more idiomatic to use `filter` on an iterator here instead of `if coeff != Scalar::zero() { ... }`

**Review by joeykraut — APPROVED** [review]
lgtm, few small suggestions

File: src/r1cs/constraint_system.rs (hunk 2 above)
Comment (akirillo) [reply src/r1cs/constraint_system.rs:34]: added a comment - this is more to handle building the `c` vector so it wouldn't make sense to call `(usize, Option<Scalar>)` a `SparseRowEntry`, that should just be a `(usize, Scalar)`

**Pull request merged**